File: src/test/java/org/stt/gui/jfx/TestFX.java
package org.stt.gui.jfx;


import com.sun.javafx.application.PlatformImpl;
import com.sun.javafx.scene.text.GlyphList;
import com.sun.javafx.scene.text.TextLayout;
import com.sun.javafx.scene.text.TextLine;
import com.sun.javafx.tk.Toolkit;
import com.sun.scenario.DelayedRunnable;
import com.sun.scenario.animation.AbstractMasterTimer;
import javafx.scene.shape.PathElement;
import javafx.scene.text.Font;

import java.io.InputStream;
import java.lang.reflect.Field;

import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.willAnswer;
import static org.mockito.Matchers.*;
import static org.mockito.Mockito.RETURNS_DEEP_STUBS;
import static org.mockito.Mockito.mock;

public class TestFX {
    private TestFX() {
    }

    /**
     * Monocle light
     */
    public static void installTK() {
        try {
            Field toolkitField = Toolkit.class.getDeclaredField("TOOLKIT");
            toolkitField.setAccessible(true);
            Toolkit toolkit = mock(Toolkit.class, RETURNS_DEEP_STUBS);

            willAnswer(invocation -> {
                Runnable run = (Runnable) invocation.getArguments()[0];
                run.run();
                return null;
            }).given(toolkit).startup(any(Runnable.class));
            given(toolkit.getMasterTimer()).willReturn(new AbstractMasterTimer() {
                @Override
                protected void postUpdateAnimationRunnable(DelayedRunnable animationRunnable) {
                }

                @Override
                protected int getPulseDuration(int precision) {
                    return 1;
                }
            });
            given(toolkit.getFontLoader().loadFont(any(InputStream.class), anyDouble())).willAnswer(invocation -> new Font(1));
            TextLayout layout = toolkit.getTextLayoutFactory().createLayout();
//            given(layout.getBounds()).willReturn(new RectBounds(0, 0, 10, 10));
            given(layout.getRuns()).willReturn(new GlyphList[0]);
            given(layout.getCaretShape(anyInt(), anyBoolean(), anyFloat(), anyFloat())).willReturn(new PathElement[0]);
            given(layout.getLines()).willReturn(new TextLine[0]);
            given(layout.getRange(anyInt(), anyInt(), anyInt(), anyFloat(), anyFloat())).willReturn(new PathElement[0]);
            toolkitField.set(Toolkit.class, toolkit);
            toolkitField.setAccessible(false);
            PlatformImpl.startup(() -> {});
        } catch (NoSuchFieldException | IllegalAccessException e) {
            throw new AssertionError(e);
        }
    }
}

File: src/test/kotlin/org/stt/gui/jfx/TestFX.kt
package org.stt.gui.jfx


import com.sun.javafx.application.PlatformImpl
import com.sun.javafx.tk.Toolkit
import com.sun.scenario.DelayedRunnable
import com.sun.scenario.animation.AbstractMasterTimer
import javafx.scene.text.Font
import org.mockito.ArgumentMatchers.*
import org.mockito.BDDMockito.given
import org.mockito.BDDMockito.willAnswer
import org.mockito.Mockito.RETURNS_DEEP_STUBS
import org.mockito.Mockito.mock
import org.stt.Matchers.any
import java.io.InputStream

object TestFX {

    /**
     * Monocle light
     */
    fun installTK() {
        try {
            val toolkitField = Toolkit::class.java.getDeclaredField("TOOLKIT")
            toolkitField.isAccessible = true
            val toolkit = mock(Toolkit::class.java, RETURNS_DEEP_STUBS)

            willAnswer { invocation ->
                val run = invocation.arguments[0] as Runnable
                run.run()
                null
            }.given(toolkit).startup(any())
            given(toolkit.masterTimer).willReturn(object : AbstractMasterTimer() {
                override fun postUpdateAnimationRunnable(animationRunnable: DelayedRunnable?) {}

                override fun getPulseDuration(precision: Int): Int {
                    return 1
                }
            })
            given(toolkit.fontLoader.loadFont(any<InputStream>(), anyDouble(), anyBoolean())).willAnswer { arrayOf( Font(1.0) ) }
            val layout = toolkit.textLayoutFactory.createLayout()
            //            given(layout.getBounds()).willReturn(new RectBounds(0, 0, 10, 10));
            given(layout.runs).willReturn(arrayOfNulls(0))
            given(layout.getCaretShape(anyInt(), anyBoolean(), anyFloat(), anyFloat())).willReturn(arrayOfNulls(0))
            given(layout.lines).willReturn(arrayOfNulls(0))
            given(layout.getRange(anyInt(), anyInt(), anyInt(), anyFloat(), anyFloat())).willReturn(arrayOfNulls(0))
            toolkitField.set(Toolkit::class.java, toolkit)
            toolkitField.isAccessible = false
            PlatformImpl.startup { }
        } catch (e: NoSuchFieldException) {
            throw AssertionError(e)
        } catch (e: IllegalAccessException) {
            throw AssertionError(e)
        }

    }
}
